File: frontend/src/api/playlists.ts
import { apiFetchJson, apiFetchVoid } from './client'
import type { Playlist, PlaylistListResponse, VideoSearchResult } from '../types'

export async function fetchPlaylists(signal?: AbortSignal) {
  return apiFetchJson<PlaylistListResponse>('/api/playlists', {
    method: 'GET',
    signal,
  })
}

export async function createPlaylist(name: string) {
  return apiFetchJson<Playlist>('/api/playlists', {
    method: 'POST',
    body: JSON.stringify({ name }),
  })
}

export async function renamePlaylist(playlistId: string, name: string) {
  return apiFetchJson<Playlist>(`/api/playlists/${playlistId}`, {
    method: 'PATCH',
    body: JSON.stringify({ name }),
  })
}

export async function deletePlaylist(playlistId: string) {
  await apiFetchVoid(`/api/playlists/${playlistId}`, {
    method: 'DELETE',
  })
}

export async function addVideoToPlaylist(
  playlistId: string,
  video: VideoSearchResult,
) {
  return apiFetchJson<Playlist>(`/api/playlists/${playlistId}/items`, {
    method: 'POST',
    body: JSON.stringify({
      video_id: video.id,
      title: video.title,
      channel_title: video.channel_title,
      thumbnail_url: video.thumbnail_url,
      source_url: video.video_url,
      duration_label: video.duration_label,
    }),
  })
}

export async function removePlaylistItem(playlistId: string, itemId: string) {
  return apiFetchJson<Playlist>(`/api/playlists/${playlistId}/items/${itemId}`, {
    method: 'DELETE',
  })
}

export async function reorderPlaylistItems(
  playlistId: string,
  orderedItemIds: string[],
) {
  return apiFetchJson<Playlist>(`/api/playlists/${playlistId}/items/reorder`, {
    method: 'PATCH',
    body: JSON.stringify({
      ordered_item_ids: orderedItemIds,
    }),
  })
}

File: backend/app/models/playlists.py
from pydantic import BaseModel, HttpUrl


class PlaylistItem(BaseModel):
    id: str
    video_id: str
    title: str
    channel_title: str
    thumbnail_url: HttpUrl | None = None
    source_url: HttpUrl
    duration_label: str | None = None
    added_at: str
    position: int


class Playlist(BaseModel):
    id: str
    name: str
    created_at: str
    updated_at: str
    items: list[PlaylistItem]


class PlaylistListResponse(BaseModel):
    items: list[Playlist]


class CreatePlaylistRequest(BaseModel):
    name: str


class UpdatePlaylistRequest(BaseModel):
    name: str


class AddPlaylistItemRequest(BaseModel):
    video_id: str
    title: str
    channel_title: str = ""
    thumbnail_url: HttpUrl | None = None
    source_url: HttpUrl
    duration_label: str | None = None


class ReorderPlaylistItemsRequest(BaseModel):
    ordered_item_ids: list[str]
